File: 다음큰숫자/java/20221223/src/Solution.java
import java.util.*;

class Solution {
    public void test() {
        System.out.println("nextNumber(78): 83 = " + nextNumber(78));
        System.out.println("nextNumber(15): 23 = " + nextNumber(15));

        System.out.println("countOne(\"1101\"): 3 = " + countOne("1101"));
    }

    public int solution(int n) {
        // test();
        int answer = nextNumber(n);
        return answer;
    }

    public int countOne(String string) {
        int count = 0;

        char[] charArray = string.toCharArray();

        for (int i = 0; i < charArray.length; i += 1) {
            if (charArray[i] == '1') {
                count += 1;
            }
        }
        return count;
    }

    public int nextNumber(int number) {
        int nextNumber = number;
        while (true) {
            nextNumber += 1;

            if (countOne(Integer.toBinaryString(number)) == countOne(Integer.toBinaryString(nextNumber))) {
                return nextNumber;
            }
        }
    }
}

File: 다음큰숫자/javascript/20221223/src/solution.js
function test() {
  console.log(`countOne('1011'): 3 = ${countOne('1011')}`);

  console.log(`nextNumber(78): 83 = ${nextNumber(78)}`);
}

function solution(n) {
  // test();
  const answer = nextNumber(n);
  return answer;
}

function countOne(string) {
  return string.split('')
    .filter((i) => i === '1')
    .length;
}

function nextNumber(number) {
  let next = number + 1;

  while (true) {
    if (countOne(number.toString(2)) === countOne(next.toString(2))) {
      return next;
    }

    next += 1;
  }
}
